### PR TITLE
[PAC][CodeGen][ELF][AArch64] Support signed GOT

### LINKED ISSUE #87984: [PAC][ELF][lld] Support signed GOT

### PR DESCRIPTION
**Pull request by kovdan01**
This re-applies #96164 after revert in #102434.

Support the following relocations and assembly operators:

- `R_AARCH64_AUTH_ADR_GOT_PAGE` (`:got_auth:` for `adrp`)
- `R_AARCH64_AUTH_LD64_GOT_LO12_NC` (`:got_auth_lo12:` for `ldr`)
- `R_AARCH64_AUTH_GOT_ADD_LO12_NC` (`:got_auth_lo12:` for `add`)

`LOADgotAUTH` pseudo-instruction is introduced which is later expanded to
actual instruction sequence like the following.

```
adrp x16, :got_auth:sym
add x16, x16, :got_auth_lo12:sym
ldr x0, [x16]
autia x0, x16
```

If a resign is requested, like below, `LOADgotPAC` pseudo is used, and GOT
load is lowered similarly to `LOADgotAUTH`.

```
@var = global i32 0
define ptr @resign_globalvar() {
  ret ptr ptrauth (ptr @var, i32 3, i64 43)
}
```

If FPAC bit is not set and auth instruction is emitted, a check+trap sequence
similar to one used for `AUT` pseudo is emitted to ensure auth success.

Both SelectionDAG and GlobalISel are suppported.
For FastISel, we fall back to SelectionDAG.

Tests starting with 'ptrauth-' have corresponding variants w/o this prefix.

See also specification
https://github.com/ARM-software/abi-aa/blob/main/pauthabielf64/pauthabielf64.rst#appendix-signed-got

### CODE REVIEW

**Comment by kovdan01** [comment]
* **#113811** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/113811?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @kovdan01 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp [hunk 1]
```diff
@@ -72,6 +72,18 @@ static bool ShouldSignWithBKey(const Function &F, const AArch64Subtarget &STI) {
   return Key == "b_key";
 }
 
+static bool HasELFSignedGOTHelper(const Function &F,
```
Comment (MaskRay) [inline llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp:75]: `hasELF...` (fix style when adding new functions)

**Review by MaskRay — APPROVED** [review]

File: llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp (hunk 1 above)
Comment (kovdan01) [reply llvm/lib/Target/AArch64/AArch64MachineFunctionInfo.cpp]: Fixed, thanks! See 48ce9aa9d4a9e50bf6cdbc75be89b8348fa4a6fb

**Comment by kovdan01** [comment]
### Merge activity

* **Nov 1, 5:17 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/113811).

**Pull request merged**